### PR TITLE
Generation doc

### PR DESCRIPTION
**Pull request by sgugger**
Add documentation (and clean docstrings) of `GenerationMixin` and `TFGenerationMixin`.

### CODE REVIEW

File: src/transformers/generation_tf_utils.py [hunk 1]
```diff
@@ -62,87 +67,76 @@ def generate(
...
+                If set to float < 1, only the most probable tokens with probabilities that add up to ``top_p`` or
+                higher are kept for generation.
+            repetition_penalty (:obj:`float`, `optional`, defaults to 1.0):
+                The parameter for repetition penalty. 1.0 means no penalty.
+            pad_token_id (:obj:`int`, `optional`):
+                The id of the `padding` token.
+            bos_token_id (:obj:`int`, `optional`):
+                The id of the `beginning-of-stream` token.
+            eos_token_id (:obj:`int`, `optional`):
+                The id of the `end-of-stream` token.
```
Comment (LysandreJik) [inline src/transformers/generation_tf_utils.py:110]: beginning-of-sequence and end-of-sequence?

File: src/transformers/generation_tf_utils.py [hunk 2]
```diff
@@ -62,87 +67,76 @@ def generate(
...
+            no_repeat_ngram_size (:obj:`int`, `optional`, defaults to 0):
+                If set to int > 0, all ngrams of that size can only occur once.
+            bad_words_ids(:obj:`List[int]`, `optional`):
+                List of token ids that are not allowed to be generated. In order to get the tokens of the words that
+                should not appear in the generated text, use :obj:`tokenizer.encode(bad_word, add_prefix_space=True)`.
+            num_return_sequences(:obj:`int`, `optional`, defaults to 1):
+                The number of independently computed returned sequences for each element in the batch.
+            attention_mask (:obj:`tf.Tensor` of :obj:`dtype=tf.int32` and shape :obj:`(batch_size, sequence_length)`, `optional`):
+                Mask to avoid performing attention on padding token indices. Mask values are in ``[0, 1]``, 1 for
+                tokens that are mot masked, and 0 for masked tokens.
```
Comment (LysandreJik) [inline src/transformers/generation_tf_utils.py:122]:
```suggestion
                tokens that are not masked, and 0 for masked tokens.
```

File: src/transformers/generation_utils.py [hunk 3]
```diff
@@ -123,89 +134,76 @@ def generate(
...
+            no_repeat_ngram_size (:obj:`int`, `optional`, defaults to 0):
+                If set to int > 0, all ngrams of that size can only occur once.
+            bad_words_ids(:obj:`List[int]`, `optional`):
+                List of token ids that are not allowed to be generated. In order to get the tokens of the words that
+                should not appear in the generated text, use :obj:`tokenizer.encode(bad_word, add_prefix_space=True)`.
+            num_return_sequences(:obj:`int`, `optional`, defaults to 1):
+                The number of independently computed returned sequences for each element in the batch.
+            attention_mask (:obj:`torch.LongTensor` of shape :obj:`(batch_size, sequence_length)`, `optional`):
+                Mask to avoid performing attention on padding token indices. Mask values are in ``[0, 1]``, 1 for
+                tokens that are mot masked, and 0 for masked tokens.
```
Comment (LysandreJik) [inline src/transformers/generation_utils.py:189]:
```suggestion
                tokens that are not masked, and 0 for masked tokens.
```

**Review by LysandreJik — APPROVED** [review]
Great, LGTM!

File: src/transformers/generation_tf_utils.py [hunk 4]
```diff
@@ -62,87 +67,76 @@ def generate(
         decoder_start_token_id=None,
         use_cache=None,
     ):
-        r""" Generates sequences for models with a LM head. The method currently supports greedy or penalized greedy decoding, sampling with top-k or nucleus sampling
-        and beam-search.
-
-        Adapted in part from `Facebook's XLM beam search code`_.
+        r"""
+        Generates sequences for models with a language modeling head. The method currently supports greedy decoding,
```
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:71]: we could mayb add here that this applies to all `AUTO_MODEL_FOR_CAUSAL_LM` and `AUTO_MODEL_FOR _SEQ2SEQ` -> not sure though whether this is very useful for the user

File: src/transformers/generation_tf_utils.py [hunk 5]
```diff
@@ -62,87 +67,76 @@ def generate(
...
+            num_beams (:obj:`int`, `optional`, defaults to 1):
+                Number of beams for beam search. 1 means no beam search.
+            temperature (:obj:`float`, `optional`, defaults tp 1.0):
+                The value used to module the next token probabilities.
+            top_k (:obj:`int`, `optional`, defaults to 50):
+                The number of highest probability vocabulary tokens to keep for top-k-filtering.
+            top_p (:obj:`float`, `optional`, defaults to 1.0):
+                If set to float < 1, only the most probable tokens with probabilities that add up to ``top_p`` or
+                higher are kept for generation.
+            repetition_penalty (:obj:`float`, `optional`, defaults to 1.0):
```
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:103]: Maybe we can link even to the original paper here, where it is explained: https://arxiv.org/pdf/1909.05858.pdf - page 5 first equation.

File: src/transformers/generation_tf_utils.py [hunk 6]
```diff
@@ -62,87 +67,76 @@ def generate(
...
+                The parameter for repetition penalty. 1.0 means no penalty.
+            pad_token_id (:obj:`int`, `optional`):
+                The id of the `padding` token.
+            bos_token_id (:obj:`int`, `optional`):
+                The id of the `beginning-of-stream` token.
+            eos_token_id (:obj:`int`, `optional`):
+                The id of the `end-of-stream` token.
+            length_penalty (:obj:`float`, `optional`, defaults to 1.0):
+                Exponential penalty to the length. 1.0 means no penalty.
```
Comment (patrickvonplaten) [inline src/transformers/generation_tf_utils.py:112]: this param can be confusing as a penalty < 1.0 penalizes length whereas a penalty > 1.0 encourages longer sequences. Maybe we can add a sentence like:
```Note: In order to encourage the model to generate shorter sequences, `length_penalty` should be set < 1.0 and to encourage longer sequences > 1.0.```

**Review by patrickvonplaten — APPROVED** [review]
Looks great! 

Would add one sentence for `length_penalty` to avoid confusion as noted down in this issue: https://github.com/huggingface/transformers/issues/4915

File: src/transformers/generation_tf_utils.py (hunk 6 above)
Comment (patrickvonplaten) [reply src/transformers/generation_tf_utils.py:112]: See https://github.com/huggingface/transformers/issues/4915

File: src/transformers/generation_tf_utils.py (hunk 4 above)
Comment (sgugger) [reply src/transformers/generation_tf_utils.py:71]: Those are not documented so it's not really useful inside the documentation.

**Pull request merged**